File: app/src/main/java/com/neuroshift/game/BouncingObstacle.java
package com.neuroshift.game;

public class BouncingObstacle extends Obstacle {

    private int screenW;

    public BouncingObstacle(float x, float y, float w, float h,
                            float vx, float vy, int color, int screenW) {
        super(x, y, w, h, vx, vy, color);
        this.screenW = screenW;
    }

    @Override
    public void update(float dt, float speedMult) {
        x += vx * speedMult * dt;
        y += vy * speedMult * dt;

        if (x < 0)             { x = 0;             vx = Math.abs(vx); }
        if (x + w > screenW)   { x = screenW - w;   vx = -Math.abs(vx); }
    }
}

File: kotlin/app/src/main/java/com/neuroshift/kotlin/BouncingObstacle.kt
package com.neuroshift.kotlin

import kotlin.math.abs

class BouncingObstacle(
    x: Float,
    y: Float,
    w: Float,
    h: Float,
    vx: Float,
    vy: Float,
    color: Int,
    private val screenW: Int
) : Obstacle(x, y, w, h, vx, vy, color) {

    override fun update(dt: Float, speedMult: Float) {
        x += vx * speedMult * dt
        y += vy * speedMult * dt

        if (x < 0f) {
            x = 0f
            vx = abs(vx)
        }
        if (x + w > screenW) {
            x = screenW.toFloat() - w
            vx = -abs(vx)
        }
    }
}
